py/binary.c

```diff
@@ -240,6 +308,16 @@ mp_obj_t mp_binary_get_val(char struct_type, char val_type, byte *p_base, byte *
         const char *s_val = (const char *)(uintptr_t)(mp_uint_t)val;
         return mp_obj_new_str(s_val, strlen(s_val));
     #if MICROPY_PY_BUILTINS_FLOAT
+    } else if (val_type == 'e') {
+        #ifdef FLT16_MAX
+        union {
+            uint16_t i;
+            _Float16 f;
+        } fpu = {val};
+        return mp_obj_new_float_from_f(fpu.f);
+        #else
+        return mp_obj_new_float_from_f(mp_decode_half(val));
+        #endif
```

Comment:
Instead of putting the "does target have float16 support" in these locations, I suggest having this line as simply:
```c
return mp_obj_new_float_from_f(mp_decode_half(val));
```
for both cases, and then logic elsewhere to provide the correct implementation of `mp_decode_half()`.  Eg:
```c
#ifdef FLT16_MAX
static inline float mp_decode_half(...) {
    union { ... }
    return fpu.f;
}
#else
...
#endif
```
This will fix a few build errors the current implementation has when float16 is available.

Reply:
I applied the above suggestion.